output/summary/data.go

```diff
@@ -0,0 +1,355 @@
+		if sample.Value == 0 {
+			atomic.AddInt64(&check.Fails, 1)
+		} else {
+			atomic.AddInt64(&check.Passes, 1)
+		}
+	}
+}
+
+// aggregatedMetricData is a container that can either hold a reference to a k6 metric stored in the registry, or
+// hold a pointer to such metric but keeping a separated Sink of values in order to keep an aggregated view of the
+// metric values. The latter is useful for tracking aggregated metric values specific to a group or scenario.
+type aggregatedMetricData map[string]aggregatedMetric
+
+// relayMetricFrom stores the metric and the metric sink from the sample. It makes the underlying metric of our
+// summary's aggregatedMetricData point directly to a metric in the k6 registry, and relies on that specific pointed
+// at metrics internal state for its computations.
+func (a aggregatedMetricData) relayMetricFrom(sample metrics.Sample) {
+	a[sample.Metric.Name] = aggregatedMetric{
+		Metric: sample.Metric,
+		Sink:   sample.Metric.Sink,
+	}
+}
+
+// addSample stores the value of the sample in a separate internal sink completely detached from the underlying metrics.
+// This allows to keep an aggregated view of the values specific to a group or scenario.
+func (a aggregatedMetricData) addSample(sample metrics.Sample) {
+	if _, exists := a[sample.Metric.Name]; !exists {
+		a[sample.Metric.Name] = newAggregatedMetric(sample.Metric)
+	}
+
+	a[sample.Metric.Name].Sink.Add(sample)
+}
+
+// FIXME (@joan): rename this to make it explicit this is different from an actual k6 metric, and this is used
+// only to keep an aggregated view of specific metric-check-group-scenario-thresholds set of values.
+type aggregatedMetric struct {
+	// FIXME (@joan): Drop this and replace it with a concrete copy of the metric data we want to track
+	// to avoid any potential confusion.
+	Metric *metrics.Metric
+
+	// FIXME (@joan): Introduce our own way of tracking thresholds, and whether they're crossed or not.
+	// Without relying on the internal submetrics the engine maintains specifically for thresholds.
+	// Thresholds []OurThreshold // { crossed: boolean }
+
+	Sink metrics.Sink
+}
+
+func newAggregatedMetric(metric *metrics.Metric) aggregatedMetric {
+	return aggregatedMetric{
+		Metric: metric,
+		Sink:   metrics.NewSink(metric.Type),
+	}
+}
+
+type aggregatedChecksData struct {
+	checks        map[string]*lib.Check
+	orderedChecks []*lib.Check
+}
+
+func newAggregatedChecksData() *aggregatedChecksData {
+	return &aggregatedChecksData{
+		checks:        make(map[string]*lib.Check),
+		orderedChecks: make([]*lib.Check, 0),
+	}
+}
+
+func (a *aggregatedChecksData) checkFor(name string) *lib.Check {
+	check, ok := a.checks[name]
+	if !ok {
+		var err error
+		check, err = lib.NewCheck(name, &lib.Group{}) // FIXME: Do we really need the group?
+		if err != nil {
+			panic(err) // This should never happen
+		}
+		a.checks[name] = check
+		a.orderedChecks = append(a.orderedChecks, check)
+	}
+	return check
+}
+
+func populateSummaryGroup(
```

Comment:
Even it's an internal, unexported method, I feel that it also maybe good to add some unit tests since the beginning, which maintain the classification and will help perhaps to cover some future cases.

Reply:
What do you think if I add a couple of tests for the:

```
func (o *Output) Summary(
	executionState *lib.ExecutionState,
	observedMetrics map[string]*metrics.Metric,
	options lib.Options,
) *lib.Summary
```

function that's public and relies on this one under the hood?

Perhaps even better? 🤔

Reply:
yep, testing of the Public API if that is possible for sure better, great idea :+1:

Reply:
Added a couple of tests in https://github.com/grafana/k6/pull/4089/commits/58e3d021ea7509bfa7aba35b0babefb7c54e04cc